src/Servers/Kestrel/Core/src/Internal/Http/HttpParser.cs

```diff
@@ -172,190 +185,132 @@ public bool ParseHeaders(TRequestHandler handler, ref SequenceReader<byte> reade
             {
-                byte ch1;
-                var ch2 = (byte)0;
-                var readAhead = 0;
+                reader.Advance(2);
+                handler.OnHeadersComplete(endStream: false);
+                return true;
+            }
 
-                // Fast path, we're still looking at the same span
-                if (span.Length >= 2)
-                {
-                    ch1 = span[0];
-                    ch2 = span[1];
-                }
-                else if (reader.TryRead(out ch1)) // Possibly split across spans
-                {
-                    // Note if we read ahead by 1 or 2 bytes
-                    readAhead = (reader.TryRead(out ch2)) ? 2 : 1;
-                }
+            var lfIndex = span.IndexOf(ByteLF);
 
-                if (ch1 == ByteCR)
-                {
-                    // Check for final CRLF.
-                    if (ch2 == ByteLF)
-                    {
-                        // If we got 2 bytes from the span directly so skip ahead 2 so that
-                        // the reader's state matches what we expect
-                        if (readAhead == 0)
-                        {
-                            reader.Advance(2);
-                        }
-
-                        // Double CRLF found, so end of headers.
-                        handler.OnHeadersComplete(endStream: false);
-                        return true;
-                    }
-                    else if (readAhead == 1)
-                    {
-                        // Didn't read 2 bytes, reset the reader so we don't consume anything
-                        reader.Rewind(1);
-                        return false;
-                    }
+            if (lfIndex > -1)
+            {
+                span = span[..lfIndex];
 
-                    Debug.Assert(readAhead == 0 || readAhead == 2);
-                    // Headers don't end in CRLF line.
+                // Consume the header including the LF
+                reader.Advance(lfIndex + 1);
+            }
 
-                    KestrelBadHttpRequestException.Throw(RequestRejectionReason.InvalidRequestHeadersNoCRLF);
-                }
+            // If the reader's span doesn't contain an LF then try to read one
+            if (lfIndex > -1 || reader.TryReadTo(out span, ByteLF, advancePastDelimiter: true))
+            {
+                // Check if there is a CR before the first LF
+                var crIndex = span.IndexOf(ByteCR);
 
-                var length = 0;
-                // We only need to look for the end if we didn't read ahead; otherwise there isn't enough in
-                // in the span to contain a header.
-                if (readAhead == 0)
+                // There is a CR, either it's right before the first LF and it's a valide CR/LF terminator
+                // or it's a lone CR and it's an invalid request
+                if (crIndex >= 0)
```

Comment:
There a some branches following, maybe it's worth to split these out to a "cold-path"?
Or just rely on PGO to do the right thing here?